[masonry] Fix intrinsic sizing with fixed-size and auto

stwrt opened the pull request:
#### a829f44a065e3bab573c8356f6e64e0f75895f37
<pre>
[masonry] Fix intrinsic sizing with fixed-size and auto
<a href="https://bugs.webkit.org/show_bug.cgi?id=277382">https://bugs.webkit.org/show_bug.cgi?id=277382</a>
<a href="https://rdar.apple.com/problem/132849745">rdar://problem/132849745</a>

Reviewed by Sammy Gill.

We were not calculating auto sizes when there is a span and it
went outside of the track sizing range. This is due to masonry
trying to place all the indefinite items in each track, and since the span would
be outside the track range it would be skipped.

The solution is to try to shift the item&apos;s start and end to fit within the track.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::shouldExcludeGridItemForMasonryTrackSizing const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/281677@main">https://commits.webkit.org/281677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eeb23ea3ce47aac0761202e48806abc9285f25c9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39959 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64531 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11147 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11438 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49022 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7745 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37222 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52497 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29850 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33913 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9740 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10060 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55777 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10027 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66260 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4544 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9898 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56390 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4565 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52471 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56563 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13508 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3770 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35764 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36846 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37939 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36591 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1436,9 +1436,15 @@ bool GridTrackSizingAlgorithm::shouldExcludeGridItemForMasonryTrackSizing(const
+            shouldExcludeGridItemForMasonryTrackSizing = true;
+        else
+            itemSpan = GridSpan::translatedDefiniteGridSpan(itemSpan.startLine() - shift, itemSpan.integerSpan() + trackIndex - shift);
+    }
```
I feel like this is not the right place to add this type of logic. Based off the current state of the function it takes in some input/constraints and answers the question of whether or not an item should be included for track sizing. I don't think we should be modifying those inputs and should instead have that be the responsibility of the caller if possible since they probably have more context as to what is going on

stwrt replied:
Fixed, moved it up to be handled by the caller.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1495,7 +1495,13 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry(GridTrack
+        unsigned shift = 0;
+        if (span.integerSpan() > 1 && ((span.integerSpan() + trackIndex) > tracks(m_direction).size())) {
+            shift = span.integerSpan() + trackIndex - tracks(m_direction).size();
+            shift = (shift > trackIndex) ? 0 : shift;
```
I think I'm having a little trouble understanding this line. What are we trying to do here that causes us to adjust `shift` to 0? (Should we even be comparing `shift` to `trackIndex`?)

stwrt replied:
It does not make sense to shift the grid item below 0. This is designed to protect against that issue. Shifting by 0 does not move the grid item at all.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1495,7 +1495,13 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry(GridTrack
 
         // Correct the span as the grid item is coming from another track.
-        span = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + span.integerSpan());
+        unsigned shift = 0;
```
```suggestion
        auto shift = 0u;
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1495,7 +1495,13 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry(GridTrack
         // Correct the span as the grid item is coming from another track.
-        span = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + span.integerSpan());
+        unsigned shift = 0;
+        if (span.integerSpan() > 1 && ((span.integerSpan() + trackIndex) > tracks(m_direction).size())) {
```
We could probably store `span.integerSpan() + trackIndex` in a local so we don't have to compute it twice and also potentially improve the readability

stwrt replied:
Good idea 👍

stwrt committed:
[masonry] Fix intrinsic sizing with fixed-size and auto
https://bugs.webkit.org/show_bug.cgi?id=277382
rdar://problem/132849745

Reviewed by Sammy Gill.

We were not calculating auto sizes when there is a span and it
went outside of the track sizing range. This is due to masonry
trying to place all the indefinite items in each track, and since the span would
be outside the track range it would be skipped.

The solution is to try to shift the item's start and end to fit within the track.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-007.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::shouldExcludeGridItemForMasonryTrackSizing const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/281677@main
